### ISSUE TITLE
Error with generic argument defaults

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#3130]
Pyright reports an error when returning an instance from generic type arguments with default values.

```python
class Base:
    pass

class Derived(Base):
    pass

BaseType = TypeVar("BaseType", bound=Base)

def instantiate(cls: Type[BaseType]) -> BaseType:
    return cls()  # No error

def instantiate_default(cls: Type[BaseType] = Derived) -> BaseType:
    # Expression of type "BaseType@instantiate_default | Derived" cannot be assigned to return type "BaseType@instantiate_default"
    #   Type "BaseType@instantiate_default | Derived" cannot be assigned to type "BaseType@instantiate_default"
    return cls()

# External usage works fine, the inferred type is "Derived", no error
foo = instantiate(Derived)
bar = instantiate_default()
```

I'm using version `1.1.225` with strict mode.

**Comment by erictraut** [comment]
In general, specifying a default argument for a parameter that is typed with a TypeVar presents a bunch of challenges for type safety. Mypy doesn't allow default arguments to be provided in this situation. Compilers in other typed languages like C# and Swift also disallow this. For more details, refer to [this thread](https://github.com/microsoft/pyright/issues/2507) and [this thread](https://github.com/python/mypy/issues/3737).

Pyright handles the situation, but it makes it safe by setting the parameter to the union of the annotation (the TypeVar) and the type of the default argument. In your example, this means the type of `cls` is `Type[BaseType | Type[Derived]`.

I'm still not 100% convinced that this is the best approach. Perhaps pyright should simply give up and disallow this case entirely like other type checkers and compilers.

A workaround I can suggest is the following:
```python
@overload
def instantiate_default() -> Derived: ...

@overload
def instantiate_default(cls: Type[BaseType]) -> BaseType: ...

def instantiate_default(cls = Derived) -> Base:
    return cls()
```

**Comment by vberlier** [comment]
I see. Thanks for the explanation.
